### PR TITLE
Algod: Minor refactoring REST client `submitForm` from go-sdk PR #335

### PR DESCRIPTION
**Pull request by ahangsu**
## Summary

The motivation of this PR is that, we want to allow for algod to simulate with option of `unlimit-log`: see #5230.

We need to send transaction with auxiliary query parameters. The current implementation in `restClient` disallowed us for such feasibility, while https://github.com/algorand/go-algorand-sdk/pull/335 contain relevant modification, and this PR mimic the changes, by separating `request` into `params` and `body`.

We expect to use this for `simulation`, where `/v2/transactions/simulate` has query parameter `unlimit-log: bool`, with raw body of txn(-group). The simulation request body resembles the one in compilation, where `/v2/teal/compile` has query parameter `sourcemap: bool` and raw body of program source bytes.

## Test Plan

Implement `Compile` with sourcemap option, and check existence of `sourcemap` field, which can be controlled by request option `Sourcemap *bool`.

### CODE REVIEW

**Review by jasonpaulos — COMMENTED** [review]
Generally looks good. This is a problem I ran into while working on simulate earlier.

In fact I made this follow-up PR to improve simulation response decoding by requesting msgpack responses: https://github.com/ahangsu/go-algorand/pull/1. I can open a PR in go-algorand for that after this one is merged

File: daemon/algod/api/client/restClient.go [hunk 1]
```diff
@@ -586,23 +586,38 @@ func (client RestClient) GetGoRoutines(ctx context.Context) (goRoutines string,
 	return
 }
 
+type compileParams struct {
+	SourceMap bool `url:"sourcemap,omitempty"`
+}
+
 // Compile compiles the given program and returned the compiled program
-func (client RestClient) Compile(program []byte) (compiledProgram []byte, programHash crypto.Digest, err error) {
+func (client RestClient) Compile(program []byte, useSourceMap bool) (compiledProgram []byte, programHash crypto.Digest, sourceMap *map[string]interface{}, err error) {
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply daemon/algod/api/client/restClient.go:594]: good point, let me do the conversion in the next few commits

**Review by jasonpaulos — DISMISSED** [review]

File: daemon/algod/api/client/restClient.go [hunk 2]
```diff
@@ -163,38 +164,52 @@ type RawResponse interface {
...
+// mergeRawQueries merges two raw queries, appending an "&" if both are non-empty
+func mergeRawQueries(q1, q2 string) string {
+	if q1 == "" {
+		return q2
+	}
+	if q2 == "" {
+		return q1
+	}
+	return q1 + "&" + q2
+}
```
Comment (bbroder-algo) [inline daemon/algod/api/client/restClient.go:176]:
```suggestion
func mergeRawQueries(q1, q2 string) string {
    if q1 == "" || q2 == "" {
        return q1 + q2
    }
    return q1 + "&" + q2
}
```

Comment (ahangsu) [reply daemon/algod/api/client/restClient.go:176]: makes sense, i take it.

File: daemon/algod/api/client/restClient.go [hunk 3]
```diff
@@ -572,23 +584,48 @@ func (client RestClient) GetGoRoutines(ctx context.Context) (goRoutines string,
...
 	if err != nil {
-		return nil, crypto.Digest{}, err
+		return nil, crypto.Digest{}, nil, err
 	}
 	programHash = crypto.Digest(progAddr)
+
+	existenceOfSourceMap := compileResponse.Sourcemap != nil
+
+	if existenceOfSourceMap != useSourceMap {
+		return nil, crypto.Digest{}, nil, fmt.Errorf("useSourceMap arg %v not agreeing with existence of source-map %v", useSourceMap, existenceOfSourceMap)
```
Comment (bbroder-algo) [inline daemon/algod/api/client/restClient.go:615]: it's an error if I don't want to use a sourcemap that I've been given?

Comment (ahangsu) [reply daemon/algod/api/client/restClient.go:615]: my bad, I think something is off with this part of logic, probably need to rewrite a bit.

Comment (ahangsu) [reply daemon/algod/api/client/restClient.go]: see [e788b1d](https://github.com/algorand/go-algorand/pull/5253/commits/e788b1da26c25fc0ef13ca0157dce09522b8319c)

**Review by bbroder-algo — APPROVED** [review]

**Review by jasonpaulos — APPROVED** [review]

**Pull request merged**